keps/1337-preempt-within-cohort-while-borrowing/README.md

```diff
@@ -0,0 +1,484 @@
+in the cohort to run both workloads at the same time.
+
+If the preemption is allowed for workloads of any priorities
+(as when `.preemption.reclaimWithinCohort` equals `Any`)
+then the following flopping happens: workload A preempts workload B, then p
+preempted workload B returns to the queue and preempts workload A to make progress.
+This repeats, and none of the workloads can make any progress.
+
+In order to mitigate this risk we only enable the preemption to reclaim
+quota within cohort, for workloads requiring borrowing when such preempted
+workloads are of lower priority. This behavior is enabled by setting
+`.preemption.borrowWithinCohort.policy: LowerPriority`.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### API
+
+```golang
+// ClusterQueuePreemption contains policies to preempt Workloads from this
+// ClusterQueue or the ClusterQueue's cohort.
+type ClusterQueuePreemption struct {
+  ...
+	// borrowWithinCohort provides configuration to allow preemption within
+	// cohort while borrowing. This configuration can only be enabled when
+	// reclaimWithinCohort is not `Never`.
+	BorrowWithinCohort *BorrowWithinCohort `json:"borrowWithinCohort,omitempty"`
+  ...
+}
+
+type BorrowWithinCohortPolicy string
+
+const (
+	BorrowWithinCohortPolicyNever         BorrowWithinCohortPolicy = "Never"
+	BorrowWithinCohortPolicyLowerPriority BorrowWithinCohortPolicy = "LowerPriority"
+)
+
+// BorrowWithinCohort contains configuration which allows to preempt workloads
+// within cohort while borrowing.
+type BorrowWithinCohort struct {
+	// policy determines the policy for preemption to reclaim quota within cohort while borrowing.
+	// Possible values are:
+	// - `Never` (default): do not reclaim quota within ClusterQueue while borrowing.
+	// - `LowerPriority`: allow preempting workloads to reclaim quota within the cohort,
+	//    but only by preempting lower-priority workloads.
+	//
+	// +kubebuilder:default=Never
+	// +kubebuilder:validation:Enum=Never;LowerPriority
+	Policy BorrowWithinCohortPolicy `json:"policy,omitempty"`
+
+	// maxPriorityThreshold allows to restrict the set of workloads for preemption
+	// to reclaim quota within the cohort, to only workloads with priority below
+	// or equal the specified level.
+	//
+	// +optional
+	MaxPriorityThreshold *int32 `json:"maxPriorityThreshold,omitempty"`
+}
+```
+
+### Implementation overview
+
+There are main modifications:
+- in the [`fitsResourceQuota`](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/scheduler/flavorassigner/flavorassigner.go#L551)
+  we return `FlavorFungibilityPolicy.Preempt` when the workload does not fit, but
```

Comment:
> we return `FlavorFungibilityPolicy.Preempt`

Does this mean `FlavorAssignmentMode=Preempt`?

Reply:
Yes, this is what I meant, let me clarify